doc/config.txt

# $Header: /Users/dossy/Desktop/cvs/aolserver/doc/Attic/config.txt,v 1.3 2000/08/09 21:45:16 kriston Exp $

#
# config.txt --
#
#      AOLserver Startup Script Reference
#
#      Nearly every AOLserver configuration variable is represented
#      here.  This file is provided as a reference -- please do not load
#      this into your server.
#


#
# Global parameters
#
ns_section "ns/parameters"
ns_param   home            $home     ;# Server's home directory

# Data structures
ns_param   dstringcachemaxentries 10 ;# Max no. of Dstrings to put on cache
ns_param   dstringcachemaxsize [expr 3*1024]
ns_param   iobufsize       16000     ;# Buffer size suitable for I/O

# Limits
ns_param   keepalivetimeout 30       ;# Max time conn is kept alive (keepalive)
                                     ;#   (set to 0 to disable keepalive)
ns_param   listenbacklog   32        ;# Max length of pending conn queue
ns_param   maxkeepalive    100       ;# Max no. of conns in keepalive state
ns_param   schedmaxelapsed 2         ;# Warn when waiting on really long procs
ns_param   shutdowntimeout 20        ;# Secs to wait on shutdown if open conns

# Server logging
ns_param   dev             true      ;# Display logging with "Dev" severity
ns_param   debug           false     ;# Display logging with "Debug" severity
ns_param   logexpanded     false     ;# true = double-spaced server.log
ns_param   logroll         false     ;# Roll server.log every 24 hours.
ns_param   maxbackup       10        ;# Max number of old server.log files
ns_param   pidfile         "nspid"   ;# PID of server (named "nspid.PORT")
ns_param   serverlog       "server.log" ;# Filename of server.log

# DNS tuning
ns_param   dnscache        true      ;# In-memory cache of DNS lookups
ns_param   dnscachetimeout 60        ;# How long to keep hostnames in cache

# Miscellaneous
ns_param   checkexitcode   false     ;# Check exit code on forked process
ns_param   crashcmd        ns_crash  ;# A Tcl command that dumps core.
ns_param   mailhost        smtp.yourcompany.com ;# SMTP host for ns_sendmail


#
# Thread library (nsthread) parameters
#
ns_section "ns/threads"
ns_param   mutexmeter      true      ;# measure lock contention
ns_param   stacksize       [expr 128*1024]  ;# stack size per thread (in bytes)


#
# MIME types.
#
# Note: AOLserver already has an exhaustive list of MIME types, but in
# case something is missing you can add it here.
#
ns_section "ns/mimetypes"
ns_param   ".xls"          "application/vnd.ms-excel"
ns_param   default         "*/*"     ;# MIME type for unknown extension
ns_param   noextension     "*/*"     ;# MIME type for missing extension


############################################################
#
# Server-level configuration
#
#  There is only one server in AOLserver, but this is helpful when multiple
#  servers share the same configuration file.  This file assumes that only
#  one server is in use so it is set at the top.
#

ns_section "ns/servers"
ns_param   $server         $servername ;# Name of virtual server.

#
# Server parameters
#
ns_section "ns/server/${servername}"
ns_param   directoryfile   $directoryfile ;# List of files to use (index.html)
ns_param   pageroot        $pageroot ;# Directory under which all pages live

# Tuning options
ns_param   connsperthread  0         ;# Normally there's one conn per thread
ns_param   flushcontent    false     ;# Flush all data before returning
ns_param   maxconnections  100       ;# Max connections to put on queue
ns_param   maxdropped      0         ;# Shut down if dropping too many conns
ns_param   maxthreads      20        ;# Tune this to scale your server
ns_param   minthreads      0         ;# Tune this to scale your server
ns_param   threadtimeout   120       ;# Idle threads die at this rate

# Limits
ns_param   maxheaders      16384     ;# Max no. of headers from client
ns_param   maxline         8192      ;# Max line length from client
ns_param   maxpost         65536     ;# Max bytes on a POST
ns_param   sendfdthreshold 2048      ;# Min size of file descriptor to send

# Performance statistics gathering
ns_param   globalstats     true      ;# Enable built-in statistics
ns_param   urlstats        true      ;# Enable URL statistics
ns_param   maxurlstats     1000      ;# Max number of URL's to do stats on

# Directory listings can be generated with an ADP or a Tcl proc.
ns_param   directoryadp    $home/dirlist.adp  ;# Choose one or the other
ns_param   directoryproc   _ns_dirlist        ;#  ...but not both!
ns_param   directorylisting simple   ;# Can be simple or fancy (for ns_dirlist)

# Miscellaneous
ns_param   checkmodifiedsince true   ;# Check url if no If-Modified-Since?
ns_param   enableaolpress  false     ;# Enable extra features used by AOLpress
ns_param   headercase      preserve  ;# tolower, toupper, preserve

#
# Internal redirects
#
ns_section "ns/server/${servername}/redirects"
ns_param   404 "/notfound.html"      ;# Not Found error page
ns_param   500 "/servererror.html"   ;# Server Error page

#
# Fastpath serves HTML
#
ns_section "ns/server/${servername}/fastpath"
ns_param   cache           true      ;# Enable cache for normal URLs
ns_param   cachemaxentry   8192      ;# Largest file size allowable in cache
ns_param   cachemaxsize    [expr 5000*1024] ;# Size of fastpath cache
ns_param   mmap            false     ;# Use mmap() for cache

#
# Tcl interpreter
#
ns_section "ns/server/${servername}/tcl"
ns_param   autoclose       true      ;# Close files the Tcl interp opened
ns_param   debug           false     ;# Names of files sourced is logged
ns_param   nsvbuckets      8         ;# No. of buckets to hold nsv's
ns_param   statlevel       0         ;# How many levels deep to run tclstats
ns_param   statmaxbuf      100       ;# Maximum entries in tclstats table
ns_param   library \
    "$home/servers/${servername}/modules/tcl" ;# Private tcl library for the server

#
# ADP (AOLserver Dynamic Page) configuration
#
ns_section "ns/server/${servername}/adp"

# ADP features
ns_param   map             "/*.adp"  ;# Extensions to parse as ADP's
ns_param   map             "/*.html" ;# Any extension can be mapped
ns_param   enableexpire    true      ;# Set "Expires: now" on all ADP's
ns_param   enabledebug     true      ;# Turn on Tclpro debugging with "?debug"
ns_param   debuginit "ns_adp_debuginit" ;# Debug procedure

# ADP tuning
ns_param   cache           true      ;# In-memory cache of ADP's
ns_param   cachesize       [expr 5000*1024]
ns_param   taglocks        false     ;# Enable post-ini creation of ADP tags
ns_param   threadcache     false     ;# default for Tcl 7.6 (nsd76)
ns_param   threadcache     true      ;# default for Tcl 8.2 (nsd82)

# ADP special pages
ns_param   errorpage \
    ${pageroot}/errorpage.adp        ;# Pretty-print ADP errors with an ADP
ns_param   startpage \
    ${pageroot}/startpage.inc        ;# Page to include for all ADP's


#
# ADP custom parsers -- see adp.c
#
ns_section "ns/server/${servername}/adp/parsers"
ns_param   adp             ".adp"   ;# The simple parser looks for <% ... %>
ns_param   fancy           ".adp"   ;# The fancy parser does a lot more.


#
# Socket driver module (HTTP)  -- nssock
#
ns_section "ns/server/${servername}/module/nssock"
ns_param   port            $httpport ;# Port for HTTP (typically 80)
ns_param   hostname        $host     ;# This is not the same as your hostname
ns_param   address         $address  ;# This is not the same as your host addr
ns_param   location        "url"     ;# URL for auto-redirects (trailing slash)

#
# Socket driver module (HTTPS) -- nsssl
#
ns_section "ns/server/${servername}/module/nsssl"
ns_param   port            $httpsport ;# Port for HTTPS (typically 443)
ns_param   hostname        $host     ;# This is not the same as your hostname
ns_param   address         $address  ;# This is not the same as your host addr
ns_param   keyfile         ${home}/servers/${servername}/modules/nsssl/keyfile.pem
ns_param   certfile        ${home}/servers/${servername}/modules/nsssl/certfile.pem


#
# Control port -- nscp
#
ns_section "ns/server/${servername}/module/nscp"
ns_param   port            9999      ;# Control port listens on port 9999
ns_param   address         "127.0.0.1" ;# For security, use 127.0.0.1 only

# Control port users
ns_section "ns/server/${servername}/module/nscp/users"
# The default password for nsadmin is "x". You should change it.
# type "ns_crypt newpassword salt" and put the encrypted string below.
ns_param user "nsadmin:t2GqvvaiIUbF2:" ;# sample user="nsadmin", pw="x".

#
# Access log -- nslog
#
ns_section "ns/server/${servername}/module/nslog"
ns_param   file            "access.log"
ns_param   formattedtime   true      ;# true=common log format
ns_param   logcombined     false     ;# true==NCSA combined format
ns_param   maxbackup       5         ;# Max number to keep around when rolling

ChangeLog
2004-06-14 20:28  Dossy Shiobara <dossy@panoptic.com>

	* nsd/connio.c: If nsend == 0, we would never call Ns_WriteConn to
	  flush the queued headers.  This could happen when we're sending
	  zero bytes of data as a response.  This fixes bug #971016 in the
	  case where fastpath.cache=false and fastpath.mmap=false and a
	  zero byte file is requested.

2004-06-14 20:25  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: unnecessary test for data != NULL actually caused
	  part of bug #971016 when fastpath.cache=false and
	  fastpath.mmap=true and the file requested is zero bytes, mmap()
	  returns 0 which gets passed along as data == NULL, causing
	  ReturnCharData() to not flush the queued headers.  removing the
	  if is safe as Ns_WriteConn will simply flush any queued data

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* Makefile: remove Makefile.module on distclean.
	* include/ns.h, nsd/log.c: Allow users to override logging functions.
	* nsd/str.c: Fix crash bug in Ns_Trim* when trimming NULL strings.

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* TAG aolserver_v40_r5

2004-03-09 13:35  Rob Crittenden <rcrittenden0569@aol.com>

	* nsd/tclhttp.c: The last header was being missed because the pointer
	  was being advanced past the \r\n.

2004-02-26 14:37  Dossy Shiobara <dossy@panoptic.com>

	* nsd/unix.c: uid_t is supposed to be unsigned, but apparently
	  isn't on OSX, apparently.

2004-02-25 19:56  Dossy Shiobara <dossy@panoptic.com>

	* nscgi/nscgi.c: Ns_ConnSetRequiredHeaders now requires length < 0
	  if you don't want a Content-Length header

2004-02-25 19:55  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: fix so that it's still possible to formulate a
	  response with no Content-Length header (such as for nscgi, or
	  HTTP streaming) where you don't know the Content-Length header in
	  advance

2004-02-24 07:15  Dossy Shiobara <dossy@panoptic.com>

	* nsd/: nsmain.c, nsd.h, driver.c: fix race condition where
	  DriverThreads are spawned but before they can Ns_SockListenEx,
	  the main thread does its NsClosePreBound and closes the prebound
	  sockets, making them unavailable to the DriverThreads if they
	  require root priv to listen on

2004-02-18 01:30  Dossy Shiobara <dossy@panoptic.com>

	* nsperm/nsperm.c: fixed bug #899364, restoring old nsperm
	  functionality as it was in 3.x.

2004-02-15  Jim Davidson <jgdavidson@aol.com>

	* nsd/conn.c:
	* include/ns.h:
	* nsd/return.c: Added new Ns_ConnGet/SetKeepAliveFlag APIs to
	allow custom request procedures to enable keep-alive directly.

	* nsd/driver.c: Minor updates to separate conn & driver structs.

2004-01-01  Scott S. Goodwin <scott@scottg.net>

        * nsd/adpparse.c: fixed bug #868600, where tag attribute values
	encapsulated within single quotes weren't being parsed properly.

2003-12-27  Scott S. Goodwin <scott@scottg.net>

        * nsd/urlopen.c: fixed bug #844809.

        * sample-config.tcl: fixed bug #805761.

2003-12-24  Scott S. Goodwin <scott@scottg.net>

	* include/nsthread.h: Bumped up NS_THREAD_NAMESIZE from 32 to 64.
	Mutex names were being truncated, making it impossible to distinguish
	mutexes with longer names when looking at stats.

        * nscgi/nscgi.c: Changed HTTPS env variable to be set to "on" or "off"
	instead of "1" or "0". There doesn't appear to be any standard for
	this, but Apache uses the on/off string in the environment.

2003-12-16  Scott S. Goodwin <scott@scottg.net>

        * nscgi/nscgi.c: Fixed issue with HTTPS environment variable not being
        set, and port not being set to 443 if connection is SSL'd. Manually
        applied these changes from a patch supplied by Vinod Kurup.

2003-11-16  Jim Davidson <jgdavidson@aol.com>

	* include/ns.h: Update version ids, added NS_SOCK_DROP and
	NS_TCL_TRACE constants and pre-queue function definitions.

	* nsd/nsd.h:  Added new "times" struct to Conn for more
	fine-grained timing.  Need to provide options to enable
	these time stamps plus ways to get the data.

	* nsd/driver.c:  Restructured in a few ways:

	1. Each loaded driver now has it's own thread.  This should
	allow some more concurrency (if necessary) and generally
	made the code more manageable.

	2. New pre-queue I/O callback API.  The API will be documented
	with an example HTTP module later.

	3. Conn allocation is handled by the driver threads and

nsd/config.c
/*
 * The contents of this file are subject to the AOLserver Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://aolserver.com/.
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is AOLserver Code and related documentation
 * distributed by AOL.
 * 
 * The Initial Developer of the Original Code is America Online,
 * Inc. Portions created by AOL are Copyright (C) 1999 America Online,
 * Inc. All Rights Reserved.
 *
 * Alternatively, the contents of this file may be used under the terms
 * of the GNU General Public License (the "GPL"), in which case the
 * provisions of GPL are applicable instead of those above.  If you wish
 * to allow use of your version of this file only under the terms of the
 * GPL and not to allow others to use your version of this file under the
 * License, indicate your decision by deleting the provisions above and
 * replace them with the notice and other provisions required by the GPL.
 * If you do not delete the provisions above, a recipient may use your
 * version of this file under either the License or the GPL.
 */

/* 
 * config.c --
 *
 *	Support for the configuration file
 */

static const char *RCSID = "@(#) $Header: /Users/dossy/Desktop/cvs/aolserver/nsd/config.c,v 1.3 2000/08/02 23:38:25 kriston Exp $, compiled: " __DATE__ " " __TIME__;

#include "nsd.h"
#define ISSLASH(c)      ((c) == '/' || (c) == '\\')

/*
 * Local procedures defined in this file.
 */

static Tcl_CmdProc SectionCmd;
static Tcl_CmdProc ParamCmd;
static Ns_Set     *GetSection(char *section, int create);
static char       *MakeSection(Ns_DString *pds, char *string);
static void        ParseConfig(char *file, char *config);
static void	   ParseAuxConfig(void);
static char       *ConfigGet(char *section, char *key,
			     int (*findproc) (Ns_Set *, char *));

/*
 * Global variables defined in this file.
 */

static Tcl_HashTable configSections;
static int initialized;


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetValue --
 *
 *	Return a config file value for a given key 
 *
 * Results:
 *	ASCIIZ ptr to a value 
 *
 * Side effects:
 *	None. 
 *
 *----------------------------------------------------------------------
 */

char *
Ns_ConfigGetValue(char *section, char *key)
{
    return ConfigGet(section, key, Ns_SetIFind);
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetValueExact --
 *
 *	Case-sensitive version of Ns_ConfigGetValue 
 *
 * Results:
 *	See Ns_ConfigGetValue 
 *
 * Side effects:
 *	None. 
 *
 *----------------------------------------------------------------------
 */

char *
Ns_ConfigGetValueExact(char *section, char *key)
{
    return ConfigGet(section, key, Ns_SetFind);
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetInt --
 *
 *	Fetch integer config values 
 *
 * Results:
 *	S_TRUE if it found an integer value; otherwise, it returns 
 *	NS_FALSE and sets the value to 0 
 *
 * Side effects:
 *	The integer value is returned by reference 
 *
 *----------------------------------------------------------------------
 */

int
Ns_ConfigGetInt(char *section, char *key, int *valuePtr)
{
    char           *s;

    s = Ns_ConfigGet(section, key);
    if (s == NULL) {
        return NS_FALSE;
    } else if (sscanf(s, "%d", valuePtr) != 1) {
        Ns_Log(Warning, "config: Ns_ConfigGetInt: "
	       "could not convert [%s]%s=\"%s\" to int",
	       section, key, s);
        return NS_FALSE;
    }

    return NS_TRUE;
}

#ifndef WIN32

/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetInt64 --
 *
 *	Like Ns_ConfigGetInt, but with INT64 data instead of 
 *	system-native int types. 
 *
 * Results:
 *	See Ns_ConfigGetInt 
 *
 * Side effects:
 *	See Ns_ConfigGetInt 
 *
 *----------------------------------------------------------------------
 */

int
Ns_ConfigGetInt64(char *section, char *key, INT64 *valuePtr)
{
    char           *s;

    s = Ns_ConfigGet(section, key);
    if (s == NULL) {
        return NS_FALSE;
    } else if (sscanf(s, NS_INT_64_FORMAT_STRING, valuePtr) != 1) {
        Ns_Log(Warning, "config: Ns_ConfigGetInt64: "
	       "could not convert [%s]%s=\"%s\" to int64",
	       section, key, s);
        return NS_FALSE;
    }

    return NS_TRUE;
}
#endif


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetBool --
 *
 *	Get a boolean config value. There are many ways to represent 
 *	a boolean value. 
 *
 * Results:
 *	NS_TRUE/NS_FALSE
 *
 * Side effects:
 *	The boolean value is returned by reference 
 *
 *----------------------------------------------------------------------
 */

int
Ns_ConfigGetBool(char *section, char *key, int *valuePtr)
{
    char           *s;

    s = Ns_ConfigGet(section, key);
    if (s == NULL) {
        return NS_FALSE;
    }
    if (STREQ(s, "1") ||
	STRIEQ(s, "y") ||
	STRIEQ(s, "yes") ||
	STRIEQ(s, "on") ||
	STRIEQ(s, "t") ||
	STRIEQ(s, "true")) {

        *valuePtr = 1;
    } else if (STREQ(s, "0") ||
	STRIEQ(s, "n") ||
	STRIEQ(s, "no") ||
	STRIEQ(s, "off") ||
	STRIEQ(s, "f") ||
	STRIEQ(s, "false")) {

        *valuePtr = 0;
    } else if (sscanf(s, "%d", valuePtr) != 1) {
        Ns_Log(Warning, "config: Ns_ConfigGetBool: "
	       "could not convert [%s]%s=\"%s\" to bool",
	       section, key, s);
        return NS_FALSE;
    }

    return NS_TRUE;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetPath --
 *
 *	Get the full name of a config file section if it exists. 
 *
 * Results:
 *	A pointer to an ASCIIZ string of the full path name, or NULL 
 *	if that path is not in the config file. 
 *
 * Side effects:
 *	None. 
 *
 *----------------------------------------------------------------------
 */

char *
Ns_ConfigGetPath(char *server, char *module, ...)
{
    va_list         ap;
    char           *s;
    Ns_DString      ds;
    Ns_Set         *setPtr;

    Ns_DStringInit(&ds);
    Ns_DStringAppend(&ds, "ns");
    if (server != NULL) {
        Ns_DStringVarAppend(&ds, "/server/", server, NULL);
    }
    if (module != NULL) {
        Ns_DStringVarAppend(&ds, "/module/", module, NULL);
    }
    va_start(ap, module);
    while ((s = va_arg(ap, char *)) != NULL) {
        Ns_DStringAppend(&ds, "/");
        while (*s != '\0' && ISSLASH(*s)) {
            ++s;
        }
        Ns_DStringAppend(&ds, s);
        while (ISSLASH(ds.string[ds.length - 1])) {
            ds.string[--ds.length] = '\0';
        }
    }
    va_end(ap);

    setPtr = Ns_ConfigSection(ds.string);
    Ns_DStringFree(&ds);

    return (setPtr ? Ns_SetName(setPtr) : NULL);
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetSections --
 *
 *	Return a malloc'ed, NULL-terminated array of sets, each 
 *	corresponding to a config section 
 *
 * Results:
 *	An array of sets 
 *
 * Side effects:
 *	The result is malloc'ed memory 
 *
 *----------------------------------------------------------------------
 */

Ns_Set **
Ns_ConfigGetSections(void)
{
    Ns_Set        **setPtrPtr;
    Tcl_HashEntry  *hPtr;
    Tcl_HashSearch  search;
    int     	    n;
    
    if (!initialized) {
	setPtrPtr = ns_calloc(1, sizeof(Ns_Set *));
    } else {
	n = configSections.numEntries + 1;
        setPtrPtr = ns_malloc(sizeof(Ns_Set *) * n);
	n = 0;
        hPtr = Tcl_FirstHashEntry(&configSections, &search);
    	while (hPtr != NULL) {
    	    setPtrPtr[n++] = Tcl_GetHashValue(hPtr);
    	    hPtr = Tcl_NextHashEntry(&search);
        }
        setPtrPtr[n] = NULL;
    }
    return setPtrPtr;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ConfigGetSection --
 *
 *	Return the Ns_Set of a config section called section. 
 *
 * Results:
 *	An Ns_Set, containing the section's parameters 
 *
 * Side effects:
 *	None. 
 *
 *----------------------------------------------------------------------
 */

Ns_Set *
Ns_ConfigGetSection(char *section)
{
    return (section ? GetSection(section, 0) : NULL);
}


/*
 *----------------------------------------------------------------------
 *
 * NsConfigRead --
 *
 *	Read a config file at startup.
 *
 * Results:
 *  	Pointer to the config buffer of an ns_malloc'ed string.
 *
 * Side Effects:
 *	Server aborts if file cannot be read for any reason.
 *
 *---------------------------------------------------------------------
 */

char *
NsConfigRead(char *file)
{
    struct stat st;
    int fd;
    char *buf;
    off_t n;

    if (stat(file, &st) != 0) {
	Ns_Fatal("config: stat(%s) failed: %s", file, strerror(errno));
    }
    if (S_ISREG(st.st_mode) == 0) {
	Ns_Fatal("config: not regular file: %s", file);
    }
    fd = open(file, O_RDONLY|O_TEXT);
    if (fd < 0) {
	Ns_Fatal("config: open(%s) failed: %s", file, strerror(errno));
    }
    n = st.st_size;
    buf = ns_malloc(n + 1);
    n = read(fd, buf, n);
    if (n < 0) {
	Ns_Fatal("config: read(%s) failed: %s", file, strerror(errno));
    }
    buf[n] = '\0';
    close(fd);
    return buf;
}


/*
 *----------------------------------------------------------------------
 *
 * NsConfigEval --
 *
 *	Eval config script in a startup Tcl interp. 
 *
 * Results:
 *  	None.
 *
 * Side Effects:
 *      Various variables in the configInterp will be set as well as
 *      the sundry configuration hashes
 *
 *---------------------------------------------------------------------
 */

void
NsConfigEval(char *config)
{
    char *err;
    Tcl_Interp *interp;
    Ns_Set     *setPtr;

    /*
     * Create an interp with a few config-related commands.
     */

    setPtr = NULL;
    interp = Tcl_CreateInterp();
    Tcl_CreateCommand(interp, "ns_log", NsTclLogCmd, NULL, NULL);
    Tcl_CreateCommand(interp, "ns_info", NsTclInfoCmd, NULL, NULL);
    Tcl_CreateCommand(interp, "ns_config", NsTclConfigCmd, NULL, NULL);
    Tcl_CreateCommand(interp, "ns_section", SectionCmd, &setPtr, NULL);
    Tcl_CreateCommand(interp, "ns_param", ParamCmd, &setPtr, NULL);

    if (Tcl_Eval(interp, config) != TCL_OK) {
	err = Tcl_GetVar(interp, "errorInfo", TCL_GLOBAL_ONLY);
	if (err == NULL) {
	    err = interp->result;
	}
	Ns_Fatal("config: script error: %s", err);
    }
    Tcl_DeleteInterp(interp);
    ParseAuxConfig();
}


/*
 *----------------------------------------------------------------------
 *
 * NsConfigParse --
 *
 *	The public interface to read the configuration file. 
 *
 * Results:
 *  	None.
 *
 * Side effects:
 *	Initial config parsed.
 *
 *----------------------------------------------------------------------
 */

void
NsConfigParse(char *file, char *config)
{
    ParseConfig(file, config);
    ParseAuxConfig();
}


/*
 *----------------------------------------------------------------------
 * ParseAuxConfig --
 *
 *	Read just the aux ini configuration files, if any.
 *
 * Results:
 *  	None.
 *
 * Side effects:
 *	Additional config may be parsed.
 *
 *----------------------------------------------------------------------
 */

static void
ParseAuxConfig(void)
{
    char        *auxdir, *config, *file;

    /*
     * Read any auxiliary config files.
     */

    auxdir = Ns_ConfigGetValue(NS_CONFIG_PARAMETERS, "auxconfigdir");
    if (auxdir != NULL) {
	struct stat     st;
	DIR            *dp;
	struct dirent  *ent;
	Ns_List        *files, *m;
	Ns_DString	ds1, ds2;
	char	       *p;

    	Ns_DStringInit(&ds1);
	Ns_DStringInit(&ds2);
	
	/*
    	 * Locate and/or verify the AuxConfigDir.
	 */
	 
	if (Ns_PathIsAbsolute(auxdir) == 0) {
    	    p = Ns_ConfigGetValue(NS_CONFIG_PARAMETERS, "home");
	    if (p == NULL) {
	    	Ns_Fatal("config: cannot locate auxconfigdir: %s", auxdir);
	    }
    	    auxdir = Ns_MakePath(&ds2, p, auxdir, NULL);
	}
	if (stat(auxdir, &st) != 0) {
	    Ns_Fatal("config: stat(%s) failed: %s", auxdir, strerror(errno));
	}
	if (S_ISDIR(st.st_mode) == 0) {
	    Ns_Fatal("config: not a directory: %s", auxdir);
	}
	
	/*
	 * Read and sort all files ending in .ini in the AuxConfigDir.
	 */
	 
	dp = opendir(auxdir);
	if (dp == NULL) {
	    Ns_Fatal("config: opendir(%s) failed: %s",
		     auxdir, strerror(errno));
	}
    	files = NULL;
	while ((ent = ns_readdir(dp)) != NULL) {

	    /*
	     * Ignore files with a dot or hash as first character.
	     */
	    if ( (ent->d_name[0] == '.')
		 || (ent->d_name[0] == '#') ) {
		continue;
	    }
	    
	    p = strrchr(ent->d_name, '.');
	    if (p != NULL && STREQ(p, ".ini")) {
		Ns_DStringTrunc(&ds1,0);
	    	Ns_MakePath(&ds1, auxdir, ent->d_name, NULL);
		Ns_ListPush(Ns_DStringExport(&ds1), files);
	    }
	}
	files = Ns_ListSort(files, (Ns_SortProc *) strcmp);
	closedir(dp);

    	/*
	 * Open and parse each .ini file.
	 */
	 
	for (m = files; m != NULL; m = Ns_ListRest(m)) {
	    file = Ns_ListFirst(m);
	    config = NsConfigRead(file);
	    ParseConfig(file, config);
	    ns_free(config);
	}
	
	Ns_ListFree(files, ns_free);
	Ns_DStringFree(&ds1);
	Ns_DStringFree(&ds2);
    }
}


/*
 *----------------------------------------------------------------------
 *
 * ConfigGet --
 *
 *	Return the value for key in the config section. 
 *
 * Results:
 *	Pointer to value, or NULL 
 *
 * Side effects:
 *	None. 
 *
 *----------------------------------------------------------------------
 */

static char *
ConfigGet(char *section, char *key, int (*findproc) (Ns_Set *, char *))
{
    Ns_Set         *setPtr;
    int             i;
    char           *s;

    s = NULL;
    if (section != NULL && key != NULL) {
        setPtr = Ns_ConfigSection(section);
        if (setPtr != NULL) {
            i = (*findproc) (setPtr, key);
            if (i >= 0) {
                s = Ns_SetValue(setPtr, i);
            }
        }
    }

    return s;
}


/*
 *----------------------------------------------------------------------
 *
 * MakeSection --
 *
 *	Make sure section and key names are trimmed and lowercase and 
 *	all backslashes (\) are converted to forward slashses (/) 
 *
 * Results:
 *	Pointer to fixed up string 
 *
 * Side effects:
 *	The string will be trimmed, lowercased, and slashes pointed 
 *	forwards. 
 *
 *----------------------------------------------------------------------
 */

static char *
MakeSection(Ns_DString *dsPtr, char *string)
{
    char           *start;
    register char  *s;

    Ns_DStringAppend(dsPtr, string);
    start = Ns_StrTrim(dsPtr->string);
    for (s = start; *s != '\0'; ++s) {
        if (*s == '\\') {
            *s = '/';
        } else if (isupper(UCHAR(*s))) {
            *s = tolower(UCHAR(*s));
        }
    }

    return start;
}


/*
 *----------------------------------------------------------------------
 *
 * ParamCmd --
 *
 *	Add a single nsd.ini parameter; this command may only be run 
 *	from within an ns_section. 
 *
 * Results:
 *	Standard Tcl Result 
 *
 * Side effects:
 *	A tcl variable will be created with the name 
 *	ns_cfgdata(section,key), and a set entry will be added for 
 *	the current section. 
 *
 *----------------------------------------------------------------------
 */

static int
ParamCmd(ClientData arg, Tcl_Interp *interp, int argc, char **argv)
{
    Ns_Set *setPtr;

    if (argc != 3) {
	Tcl_AppendResult(interp, "wrong # args: should be \"",
		argv[0], " key value", NULL);
	return TCL_ERROR;
    }
    setPtr = *((Ns_Set **) arg);
    if (setPtr == NULL) {
	Tcl_AppendResult(interp, argv[0],
			 " not preceded by an ns_section command.", NULL);
	return TCL_ERROR;
    }
    Ns_SetPut(setPtr, argv[1], argv[2]);

    return TCL_OK;
}


/*
 *----------------------------------------------------------------------
 *
 * SectionCmd --
 *
 *	This creates a new config section and sets a shared variable
 *	to point at a newly-allocated set for holding config data.
 *	ns_param stores config data in the set.
 *
 * Results:
 *	Standard tcl result. 
 *
 * Side effects:
 *	Section set is created (if necessary).
 *
 *----------------------------------------------------------------------
 */

static int
SectionCmd(ClientData arg, Tcl_Interp *interp, int argc, char **argv)
{
    Ns_Set  **setPtrPtr;

    if (argc != 2) {
	Tcl_AppendResult(interp, "wrong # args: should be \"",
			 argv[0], " sectionname", NULL);
	return TCL_ERROR;
    }
    setPtrPtr = (Ns_Set **) arg;
    *setPtrPtr = GetSection(argv[1], 1);

    return TCL_OK;
}


/*
 *----------------------------------------------------------------------
 *
 * GetSection --
 *
 *	Creates and/or gets a config section.
 *
 * Results:
 *	Pointer to new or existing Ns_Set for given section.
 *
 * Side effects:
 *	Section set created (if necessary).
 *
 *----------------------------------------------------------------------
 */

static Ns_Set *
GetSection(char *section, int create)
{
    Ns_DString ds;
    Tcl_HashEntry *hPtr;
    int new;
    Ns_Set *setPtr;

    if (!initialized) {
	Tcl_InitHashTable(&configSections, TCL_STRING_KEYS);
	initialized = 1;
    }

    Ns_DStringInit(&ds);
    setPtr = NULL;
    MakeSection(&ds, section);    
    if (!create) {
	hPtr = Tcl_FindHashEntry(&configSections, ds.string);
    } else {
    	hPtr = Tcl_CreateHashEntry(&configSections, ds.string, &new);
    	if (new) {
	    setPtr = Ns_SetCreate(section);
	    Tcl_SetHashValue(hPtr, setPtr);
	}
    }
    if (hPtr != NULL) {
	setPtr = Tcl_GetHashValue(hPtr);
    }
    Ns_DStringFree(&ds);
    return setPtr;
}


/*
 *----------------------------------------------------------------------
 *
 * ParseConfig --
 *
 *	Parse config lines in .ini format.
 *
 * Results:
 *	None 
 *
 * Side effects:
 *	Ns_Sets will be created for each section and linked to from 
 *	the configSections hash table.
 *
 *----------------------------------------------------------------------
 */

static void
ParseConfig(char *file, char *config)
{
    Ns_DString	    ds;
    Ns_Set         *setPtr;
    char    	   *line, *nextLine, *s, *e;

    Ns_DStringInit(&ds);
    setPtr = NULL;

    for (line = config; line != NULL; line = nextLine) {

    	/*
	 * Locate and trim next line.
	 */
	 
	nextLine = strchr(line, '\n');
	if (nextLine != NULL) {
	    *nextLine++ = '\0';
	}
	line = Ns_StrTrim(line);

	switch (*line) {
	    case ';':
	    case '\0':

		/*
		 * Skip comments and blank lines.
		 */

		break;

	    case '[':

		/*
		 * Parse section name and begin new section.
		 */
		 
		s = line + 1;
		e = strchr(s, ']');
		if (e == NULL) {
		    Ns_Log(Warning, "config: ParseConfig: "
			   "invalid section name: %s", line);
		} else {
		    *e = '\0';
		    if (*s == '\0') {
			Ns_Log(Warning, "config: ParseConfig: "
			       "null section name");
		    } else {
			setPtr = GetSection(s, 1);
		    }
		}
		break;

	    default:

		/*
		 * Parse a section entry.
		 */
		 
		if (setPtr == NULL) {
		    Ns_Log(Warning, "config: ParseConfig: "
			   "entry before section: %s", line);
		} else {
		    e = strchr(line, '=');
		    if (e == NULL) {
			Ns_Log(Warning, "config: ParseConfig: "
			       "invalid entry: %s", line);
		    } else {
			*e++ = '\0';
			e = Ns_StrTrim(e);
			s = Ns_StrTrim(line);
			Ns_SetPut(setPtr, s, e);
		    }
		}
		break;
	}
    }
    Ns_DStringFree(&ds);
}

